io_scene_niftools/modules/nif_import/geometry/vertex/__init__.py

```diff
@@ -67,19 +68,17 @@ def map_normals(b_mesh, n_tri_data):
         assert len(b_mesh.vertices) == len(n_tri_data.normals)
         # set normals
         if NifOp.props.use_custom_normals:
-            # map normals so we can set them to the edge corners (stored per loop)
-            no_array = []
-            for face in b_mesh.polygons:
-                for vertex_index in face.vertices:
-                    # no_array.append(model.normals[vertex_index])
-                    # no_array.append(mathutils.Vector(n_tri_data.normals[vertex_index]).normalized())
-                    no_array.append(n_tri_data.normals[vertex_index].as_tuple())
-                    # no_array.append((0,0,1))
-                # no_array.append(model.tangents[vertex_index])
-                # face.use_smooth = True
+            no_array = np.array([normal.as_tuple() for normal in n_tri_data.normals])
+            # the normals need to be pre-normalized or blender will do it inconsistely, leading to marked sharp edges
+            no_norms = np.linalg.norm(no_array, ord=2, axis=1, keepdims=True)
+            non_zero_norms = np.reshape(no_norms != 0, newshape=len(no_norms))
```

Comment:
Not really, but I thought better safe than sorry. Plus, if someone does have it, it can be a pain for them to fix. Should the separate normalize() function be in a different file? After all, normalization is not unique to this one application.

Reply:
Absolutely.
Do we need it elsewhere? If we do eventually we can pull it out to the math helpers file. For now just an extracted method is fine.